### PR TITLE
feat: update camper model

### PR DESCRIPTION
**Pull request by NirmalHegde**
## Notion ticket link

[Ticket Name](https://uwblueprintexecs.notion.site/Camper-model-updates-4269f81c5d8d418f84b86de904a0f815)

## Implementation description
* Added new optionalClauses element to camper model
* Edited earlyDropoff and latePickup to accommodate daily changes in timings

## Steps to test
1. ensure camper model is correct and to the specifics of the required design

## What should reviewers focus on?
* camper.model.ts
* camperService.ts

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/types.ts [hunk 1]
```diff
@@ -38,8 +38,8 @@ export type CamperDTO = {
   allergies: string;
   hasCamera: boolean;
   hasLaptop: boolean;
-  earlyDropoff: string;
-  latePickup: string;
+  earlyDropoff: Date[];
```
Comment (jjjding2) [inline backend/typescript/types.ts:41]: I don't believe Date objects are supported for DTOs. We should use strings and rely on validators to ensure they are in the proper format

File: backend/typescript/middlewares/validators/camperValidators.ts [hunk 2]
```diff
@@ -185,23 +216,33 @@ export const updateCamperDtoValidator = async (
...
-      !validateTime(req.body.latePickup))
-  ) {
-    return res
-      .status(400)
-      .send(getApiValidationError("latePickup", "24 hr time string"));
+  // eslint-disable-next-line no-restricted-syntax
+  for (const dropoffDate of req.body.earlyDropoff) {
+    if (
+      !validatePrimitive(dropoffDate, "Date string") ||
+      !validateTime(dropoffDate)
```
Comment (jjjding2) [inline backend/typescript/middlewares/validators/camperValidators.ts:226]: Should we be using validateDate here? (i.e. we aren't storing times like "12:30")

**Review by jjjding2 — CHANGES_REQUESTED** [review]
Some validator / typing comments

File: backend/typescript/types.ts (hunk 1 above)
Comment (NirmalHegde) [reply backend/typescript/types.ts:41]: registration date is also a Date format, so I thought it was ok. Lmk!

Comment (jjjding2) [reply backend/typescript/types.ts:41]: Hmm ok, if it works in testing maybe it's fine because of type conversions

**Review by annechung — CHANGES_REQUESTED** [review]

File: backend/typescript/models/camper.model.ts [hunk 3]
```diff
@@ -9,8 +9,8 @@ export interface Camper extends Document {
   allergies: string;
   hasCamera: boolean;
   hasLaptop: boolean;
-  earlyDropoff: string;
-  latePickup: string;
+  earlyDropoff: string[];
+  latePickup: string[];
```
Comment (annechung) [inline backend/typescript/models/camper.model.ts:13]: sorry I should have clarified, these should be Date types in the model itself but just strings in the DTOs D:

File: backend/typescript/models/camper.model.ts [hunk 4]
```diff
@@ -19,14 +19,20 @@ export interface Camper extends Document {
     phoneNumber: string;
   }[];
   formResponses: Map<string, string>;
-  registrationDate: Date;
+  registrationDate: string;
```
Comment (annechung) [inline backend/typescript/models/camper.model.ts:22]:
```suggestion
  registrationDate: Date;
```

File: backend/typescript/middlewares/validators/camperValidators.ts [hunk 5]
```diff
@@ -55,23 +54,38 @@ export const createCampersDtoValidator = async (
...
-      return res
-        .status(400)
-        .send(getApiValidationError("latePickup", "24 hr time string"));
+    if (!Array.isArray(camper.earlyDropoff)) {
+      return res.status(400).send("early dropoff is not an array");
+    }
+    // eslint-disable-next-line no-restricted-syntax
+    for (const dropoffDate of camper.earlyDropoff) {
+      if (
+        !validatePrimitive(dropoffDate, "Date string") ||
```
Comment (annechung) [inline backend/typescript/middlewares/validators/camperValidators.ts:68]:
```suggestion
        !validatePrimitive(dropoffDate, "string") ||
```

File: backend/typescript/middlewares/validators/camperValidators.ts [hunk 6]
```diff
@@ -55,23 +54,38 @@ export const createCampersDtoValidator = async (
...
+          .send(getApiValidationError("earlyDropoff", "Date string"));
+      }
+    }
+    if (!Array.isArray(camper.latePickup)) {
+      return res.status(400).send("late pickup is not an array");
+    }
+    // eslint-disable-next-line no-restricted-syntax
+    for (const pickupDate of camper.latePickup) {
+      if (
+        !validatePrimitive(pickupDate, "Date string") ||
```
Comment (annechung) [inline backend/typescript/middlewares/validators/camperValidators.ts:82]:
```suggestion
        !validatePrimitive(pickupDate, "string") ||
```

**Review by annechung — APPROVED** [review]
looks gouda 🧀

File: backend/typescript/middlewares/validators/camperValidators.ts [hunk 7]
```diff
@@ -55,23 +54,38 @@ export const createCampersDtoValidator = async (
         .status(400)
         .send(getApiValidationError("hasLaptop", "boolean"));
     }
-    if (
-      camper.earlyDropoff &&
-      (!validatePrimitive(camper.earlyDropoff, "string") ||
-        !validateTime(camper.earlyDropoff))
-    ) {
+    if (!Array.isArray(camper.contacts) || camper.contacts.length !== 2) {
```
Comment (annechung) [inline backend/typescript/middlewares/validators/camperValidators.ts:57]: can you delete this check?? just realized this was duplicated

**Review by jjjding2 — APPROVED** [review]

**Pull request merged**